### PR TITLE
Named indices

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3380/

**Comment by danielhlarkin** [comment]
Documentation still pending.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3380/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3393/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3394/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3416/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3421/

File: arangod/Indexes/Index.h [hunk 1]
```diff
@@ -113,6 +113,17 @@ class Index {
   /// @brief return the index id
   inline TRI_idx_iid_t id() const { return _iid; }
 
+  /// @brief return the index name
+  inline std::string const& name() const {
+    if (_name == StaticStrings::IndexNameEdgeFrom || _name == StaticStrings::IndexNameEdgeTo) {
+      return StaticStrings::IndexNameEdge;
```
Comment (jsteemann) [inline arangod/Indexes/Index.h]: Is this going to cause problems if two indexes have the same name, e.g. in the RocksDB engine?
Additionally, what happens if a user creates an index with name `StaticStrings::IndexNameEdge` in a non-edge collection?

File: arangod/Indexes/IndexFactory.cpp [hunk 2]
```diff
@@ -220,6 +219,29 @@ Result IndexFactory::enhanceIndexDefinition( // normalizze deefinition
...
+    std::string name;
+
+    if (nameSlice.isString() && (nameSlice.getStringLength() != 0)) {
+      name = nameSlice.copyString();
+    } else {
+      // we should set the name for special types explicitly elsewhere, but just in case...
+      if (Index::type(type.copyString()) == Index::IndexType::TRI_IDX_TYPE_PRIMARY_INDEX) {
+        name = StaticStrings::IndexNamePrimary;
+      } else if (Index::type(type.copyString()) == Index::IndexType::TRI_IDX_TYPE_EDGE_INDEX) {
+        name = StaticStrings::IndexNameEdge;
```
Comment (jsteemann) [inline arangod/Indexes/IndexFactory.cpp]: Will this lead to two edge indexes (in the RocksDB engine) end up with the same name? Is this going to be a problem?

File: arangod/Indexes/Index.h (hunk 1 above)
Comment (jsteemann) [reply arangod/Indexes/Index.h]: And how does one distinguish between the `_from` part of the `_to` part in the RocksDB engine when running a query and using the "force index" feature? If both edge indexes pretend to have the same name, as a user there is no way to select the "right" one of the two, right?

File: tests/js/common/shell/shell-index-ensure.js [hunk 3]
```diff
@@ -107,6 +112,76 @@ function ensureIndexSuite() {
...
+      assertEqual("skiplist", idx.type);
+      assertFalse(idx.unique);
+      assertEqual([ "b", "d" ], idx.fields);
+      assertEqual(name, idx.name);
+
+      // expect duplicate name to fail and return old index with same name
+      idx = collection.ensureIndex({ type: "hash", fields: [ "a", "c" ], name: name });
+      assertEqual("skiplist", idx.type);
+      assertFalse(idx.unique);
+      assertEqual([ "b", "d" ], idx.fields);
```
Comment (jsteemann) [inline tests/js/common/shell/shell-index-ensure.js]: Is this actually the expected behavior?
If the index is created with a different definition than an existing index but using a name of an already existing index, I think the request should fail. This is of course debatable, but returning an index on different fields I find somewhat unexpected.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3424/

File: arangod/Indexes/Index.cpp [hunk 4]
```diff
@@ -413,18 +413,28 @@ bool Index::validateHandle(char const* key, size_t* split) {
...
+    return true;
+  }
+
+  return false;
+}
+
 /// @brief index comparator, used by the coordinator to detect if two index
 /// contents are the same
 bool Index::Compare(VPackSlice const& lhs, VPackSlice const& rhs) {
-  {
```
Comment (jsteemann) [inline arangod/Indexes/Index.cpp]: This is a public static method and is also called from here:
```
arangod/Cluster/Maintenance.cpp:              if (!arangodb::Index::Compare(pindex, lindex)) {
arangod/ClusterEngine/ClusterIndex.cpp:  return Index::Compare(_info.slice(), info);
```
Do the above place need adjustment too?

Comment (danielhlarkin) [reply arangod/Indexes/Index.cpp]: Nope, these are okay to use the old check.

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3446/

**Pull request merged**